### PR TITLE
Introduce Model building

### PR DESCRIPTION
**Pull request by Morriar**
While we're working on making the RubyIndexer a standalone gem, we can start making room for it in Spoom so we can base the dead code analysis.

In this PR we create a representation close to what the RubyIndexer builds so we'll be able to use it in dead code analysis in subsequent PRs.

This will allow us to test the concepts and see what's missing to the RubyIndexer API to build other tools than an LSP server.

### CODE REVIEW

File: lib/spoom/model/namespace_visitor.rb [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+          end
+        else
+          super
+        end
+      end
+
+      private
+
+      sig { returns(T::Array[String]) }
+      attr_reader :names_nesting
```
[the remark this answers is not in the document]
Comment (Morriar) [reply lib/spoom/model/namespace_visitor.rb:52]: It's used by the subclasses: https://github.com/Shopify/spoom/pull/557/files#diff-983e178c03c8a9daa368e28fea30cbd7c71064568c9d9ff47e0b2bfe03b754beR24

[the remark this answers is not in the document]
Comment (Morriar) [reply lib/spoom/model/namespace_visitor.rb:52]: Actually, I can remove the accessor and use the instance var directly 👍

File: lib/spoom/model/builder.rb [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+        else
+          @last_sigs.clear
+          super
+        end
+      end
+
+      private
+
+      sig { returns(T::Array[Sig]) }
+      def collect_sigs
```
[the remark this answers is not in the document]
Comment (Morriar) [reply lib/spoom/model/builder.rb:212]: For code like this:

```rb
sig { void }
sig { returns(Integer) }
def foo; end
```

As we visit the tree, we accumulate the signatures into the `last_sigs` array.

When we get to actually building the `Method`, we "collect" all the last sigs and reset the array.

**Review by vinistock — APPROVED** [review]
I'm so excited to get Spoom on the indexer. It's the culmination of a lot of work!

**Pull request merged**